akka-javasdk/src/main/scala/akka/javasdk/impl/agent/ChatAgentEffectImpl.scala

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.javasdk.impl.agent
+
+import akka.annotation.InternalApi
+import akka.javasdk.Metadata
+import akka.javasdk.agent.ChatAgent.Effect
+import akka.javasdk.agent.ChatAgent.Effect.Builder
+import akka.javasdk.agent.ChatAgent.Effect.OnSuccessBuilder
+import akka.javasdk.impl.effect.ErrorReplyImpl
+import akka.javasdk.impl.effect.MessageReplyImpl
+import akka.javasdk.impl.effect.NoSecondaryEffectImpl
+import akka.javasdk.impl.effect.SecondaryEffectImpl
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[javasdk] object ChatAgentEffectImpl {
+  sealed trait PrimaryEffectImpl
+  object RequestModel {
+    val empty: RequestModel =
+      RequestModel(systemMessage = "", userMessage = "", responseType = classOf[String], replyMetadata = Metadata.EMPTY)
+  }
+  final case class RequestModel(
+      systemMessage: String,
+      userMessage: String,
+      responseType: Class[_],
+      replyMetadata: Metadata)
+      extends PrimaryEffectImpl
+  case object NoPrimaryEffect extends PrimaryEffectImpl
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[javasdk] final class ChatAgentEffectImpl[Reply] extends Builder with OnSuccessBuilder with Effect[Reply] {
+  import ChatAgentEffectImpl._
+
+  private var _primaryEffect: PrimaryEffectImpl = NoPrimaryEffect
+  private var _secondaryEffect: SecondaryEffectImpl = NoSecondaryEffectImpl
+
+  def primaryEffect: PrimaryEffectImpl = _primaryEffect
+
+  def secondaryEffect: SecondaryEffectImpl =
+    _secondaryEffect
+
+  override def reply[T](message: T): ChatAgentEffectImpl[T] =
+    reply(message, Metadata.EMPTY)
+
+  override def reply[T](message: T, metadata: Metadata): ChatAgentEffectImpl[T] = {
+    _secondaryEffect = MessageReplyImpl(message, metadata)
+    this.asInstanceOf[ChatAgentEffectImpl[T]]
+  }
+
+  override def error[T](description: String): ChatAgentEffectImpl[T] = {
+    _secondaryEffect = ErrorReplyImpl(description)
+    this.asInstanceOf[ChatAgentEffectImpl[T]]
+  }
+
+  def hasError(): Boolean =
+    _secondaryEffect.isInstanceOf[ErrorReplyImpl]
+
+  override def systemMessage(message: String): Builder = {
```

Comment:
I see a potential challenge here, if we introduce a method `systemMessageFromTemplate`, then in the implementation, I would like to use the component client to get the template value, but this is Scala code, so I can't use method reference anymore.

Reply:
I could use `EntityClient` instead, maybe that's good enough.

Reply:
Ok, I think that the actual call should be moved to `ChatAgentImpl`, so we should extend RequestModel to:
```
final case class RequestModel(
      systemMessage: String,
      systemTempleteId: String,
      userMessage: String,
      responseType: Class[_],
      replyMetadata: Metadata)
```
and some null checking later, or better with ADT like:
```
trait SystemMessage

case class ConstantSystemMessage(...)
case class TempleteSystemMessage().


final case class RequestModel(
      systemMessage: SystemMessage,
      userMessage: String,
      responseType: Class[_],
      replyMetadata: Metadata)
```

Reply:
We can always call out to an internal java method to get a hold of the method handle.